### PR TITLE
[AIRFLOW-591] Contribution: Integration with datadog

### PR DESCRIPTION
**Pull request by gtoonstra**
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
https://issues.apache.org/jira/browse/AIRFLOW-591

Testing Done:
- A unit test is in place for the sensor, which also tests parts of the hook. I system tested the remainder, because a remote service setup is needed to really test this end-to-end.

Looking forward to your comments.

### CODE REVIEW

**Comment by r39132** [comment]
Interesting.. we use DataDog.. will review and test the PR and get back to you soon.

**Comment by msumit** [comment]
+1 for idea.. we also use datadog..

**Comment by alexvanboxel** [comment]
+1 we're evaluating dog, I'll have a talk with our service manager if he likes to try it.

**Comment by alexvanboxel** [comment]
Maybe add a comment of the extra dependency that needs to be included.

File: airflow/contrib/hooks/datadog_hook.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            'app_key': self.app_key
+        }
+        initialize(**options)
+
+    def validate_response(self, response):
+        if response['status'] != 'ok':
+            logging.error("Data dog returned: "+response)
+            raise AirflowException("Error status received from datadog")
+
+    """
```
Comment (skudriashev) [inline airflow/contrib/hooks/datadog_hook.py]: Looks like this is wrong place for the docstring.

File: airflow/contrib/hooks/datadog_hook.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        response = api.Metric.send(
+            metric=metric_name,
+            points=datapoint,
+            host=self.host,
+            tags=tags)
+
+        self.validate_response(response)
+        return response
+
+    """
```
Comment (skudriashev) [inline airflow/contrib/hooks/datadog_hook.py]: Same here.

File: airflow/contrib/hooks/datadog_hook.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    def query_metric(self,
+                     query,
+                     from_seconds_ago,
+                     to_seconds_ago):
+        now = int(time.time())
+
+        response = api.Metric.query(
+            start=now - from_seconds_ago,
+            end=now-to_seconds_ago,
```
Comment (skudriashev) [inline airflow/contrib/hooks/datadog_hook.py]: Whitespaces around `-` (pep8)

File: airflow/contrib/sensors/datadog_sensor.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        if isinstance(response, dict) and response.get('status', 'ok') != 'ok':
+            logging.error("Unexpected datadog result: %s" % (response))
+            raise AirflowException("Datadog returned unexpected result")
+
+        if self.response_check:
+            # run content check on response
+            return self.response_check(response)
+
+        # If no check was inserted, assume any event that matched yields true.
+        if len(response) > 0:
```
Comment (skudriashev) [inline airflow/contrib/sensors/datadog_sensor.py]: Just:

``` python
    return len(response) > 0
```

**Review by skudriashev — CHANGES_REQUESTED** [review]

**Comment by gtoonstra** [comment]
I fixed the obvious docstrings. I looked into the whitespaces around - (pep8), but when these are parameters it's supposed to be like this. I ran flake8 and it started complaining when I inserted the spaces.

**Comment by skudriashev** [comment]
@gtoonstra,

> I fixed the obvious docstrings. I looked into the whitespaces around - (pep8), but when these are parameters it's supposed to be like this. I ran flake8 and it started complaining when I inserted the spaces.

Just make it consistent with previous line.

**Comment by gtoonstra** [comment]
@skudriashev : I tried doing that, but PEP8 prescribes to not use spaces when the equals sign is used as a keyword argument or default parameter value: 

> Don't use spaces around the = sign when used to indicate a keyword argument or a default parameter value.

The first one is a basic assignment. I'm not sure if one of the options of PEP8 is turned off?

E251 unexpected spaces around keyword / parameter equals

If that warning is turned off, I'll make the change and move forward.

**Comment by skudriashev** [comment]
@gtoonstra, I've tried locally:

``` bash
$ pip install flake8
$ pip freeze
...
flake8==3.0.4
```

``` bash
$ flake8 airflow/contrib/hooks/datadog_hook.py
...
tmp.py:58:48: E226 missing whitespace around arithmetic operator                                                                                                                                                                             
tmp.py:98:20: E226 missing whitespace around arithmetic operator
...
```

Fixing this two lines make linting pass fine.

**Comment by gtoonstra** [comment]
Allright, flake8 wasn't failing for me because since 2012 E226 is ignored by default. I had to enable this with --select=E226 to get flake8 to complain. Probably you've got this configured on by default. Should be fixed now.

**Comment by r39132** [comment]
@gtoonstra can you go ahead and update setup.py with the new dependency (i.e. datadog). You can list it in the extras_require section, such as we do for packages like s3 and postgres, so that you can do `pip install "airflow[s3, postgres]"`

**Comment by r39132** [comment]
Also, please squash your commits.

**Comment by gtoonstra** [comment]
Changes made, commits squashed.

**Comment by r39132** [comment]
+1

**Pull request closed without merging**